src/users/users.service.ts

```diff
@@ -60,4 +62,19 @@ export class UsersService {
     }
     return user;
   }
+
+  async resetPassword(userId: number, resetPasswordDto: ResetPasswordDto): Promise<void> {
+    const user = await this.userRepository.findOneOrFail(userId);
+
+    const isOldPasswordValid = await bcrypt.compare(resetPasswordDto.oldPassword, user.password);
+    if (!isOldPasswordValid) {
+      throw new UnauthorizedException("Old password is incorrect");
+    }
+
+    if (resetPasswordDto.oldPassword === resetPasswordDto.newPassword) {
+      throw new BadRequestException("New password must be different from the old password");
+    }
```

Comment:
I think the error should be a bit more vague, maybe something like "cannot set a previously used password"